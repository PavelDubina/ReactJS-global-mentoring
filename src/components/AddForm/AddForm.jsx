import React, { useState } from 'react'
import { useDispatch } from 'react-redux'
import PropTypes from 'prop-types'
import { Button } from '../Button/Button'
import { FormSelect } from '../FormSelect/FormSelect'
import { Modal } from '../Modal/Modal'
import { Input } from '../Input/Input'
import addMovie from '../../redux/actions/addMovie'
import styles from './AddForm.scss'

export const AddForm = ({ onClose }) => {
  const dispatch = useDispatch()
  const initialState = {
    title: '',
    release_date: '',
    overview: '',
    runtime: '',
    genres: [],
    poster_path: '',
  }
  const [state, setState] = useState(initialState)
  const { title, release_date, overview, runtime, genres, poster_path } = state
  const handleInput = (event) => {
    const { value, name } = event.target
    setState({ ...state, [name]: name === 'runtime' ? parseInt(value, 10) : value })
  }
  const handleGenre = (event) => {
    const { value, checked } = event.target
    return checked && !genres.includes(value)
      ? setState((state) => ({ ...state, genres: [...state.genres, value] }))
      : setState((state) => ({ ...state, genres: state.genres.filter((genre) => genre !== value) }))
  }
  const handleReset = () => setState(initialState)
  const onSubmit = () => dispatch(addMovie(state))
  return (
    <Modal title="ADD MOVIE" onClose={onClose}>
      <form className={styles.form}>
        <Input title="TITLE" name="title" type="text" value={title} placeholder="Title here" onChange={handleInput} />
        <br />
        <Input
          title="RELEASE DATE"
          name="release_date"
          type="date"
          value={release_date}
          style={{ color: release_date ? '#fff' : '#a9a9a9c0' }}
          onChange={handleInput}
        />
        <br />
        <Input
          title="POSTER PATH"
          name="poster_path"
          type="text"
          value={poster_path}
          placeholder="Poster URL here"
          onChange={handleInput}
        />
        <br />
        <FormSelect value={genres.join(', ')} onChange={handleGenre} />
        <br />
        <Input
          title="OVERVIEW"
          name="overview"
          type="text"
          value={overview}
          placeholder="Overview here"
          onChange={handleInput}
        />
        <br />
        <Input
          title="RUNTIME"
          name="runtime"
          type="text"
          value={runtime}
          placeholder="Runtime here"
          onChange={handleInput}
        />
      </form>
      <div className={styles.btns}>
        <Button onClick={handleReset} styleType="reset">
          RESET
        </Button>
        <Button isClose onClick={onSubmit} styleType="confirm">
          SUBMIT
        </Button>
      </div>
    </Modal>
  )
}

AddForm.propTypes = {
  onClose: PropTypes.func.isRequired,
}
