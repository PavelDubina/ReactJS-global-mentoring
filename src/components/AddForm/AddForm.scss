.form {
  padding: 16px 0;
  display: flex;
  flex-direction: column;
  color: #f65261;

  & input[type='date'] {
    color: #a9a9a9c0;
  }

  & input[type='date']::-webkit-calendar-picker-indicator {
    cursor: pointer;
    opacity: 0.6;
    outline: none;
    filter: invert(1);

    &:hover {
      opacity: 1;
    }
  }
}

.label {
  display: flex;
  flex-direction: column;
}

.input {
  margin: 8px 0 0 0;
  padding: 0 16px;
  background-color: #232323c9;
  border-radius: 5px;
  border-width: 0;
  font-size: 24px;
  line-height: 48px;
  color: #fff;
  outline: none;
}

.btns {
  display: flex;
  justify-content: flex-end;
  margin: 64px 0 0 0;

  & button[type='reset'] {
    margin: 0 16px 0 0;
  }
}
