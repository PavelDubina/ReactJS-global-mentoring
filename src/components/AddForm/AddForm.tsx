import React from 'react'
import { useDispatch } from 'react-redux'
import { Formik } from 'formik'
import * as yup from 'yup'
import { useTranslation } from 'next-i18next'
import { Button, EStyleTypeBtn, EType } from '../Button/Button'
import { FormSelect } from '../FormSelect/FormSelect'
import { Modal } from '../Modal/Modal'
import { Input } from '../Input/Input'
import addMovie from '../../redux/actions/addMovie'
import styles from '../../../styles/AddForm.module.scss'

type AddFormProps = {
  onClose: (isMessage?: boolean) => void
}

// PATTERN:{Destructuring props}
export const AddForm: React.FC<AddFormProps> = ({ onClose }) => {
  const { t } = useTranslation()
  const dispatch = useDispatch()
  const validationSchema = yup.object().shape({
    title: yup.string().required(t('addModal.validation.title')),
    poster_path: yup
      .string()
      .required(t('addModal.validation.poster.required'))
      .url(t('addModal.validation.poster.url')),
    release_date: yup
      .date()
      .max(new Date(), t('addModal.validation.date.max'))
      .required(t('addModal.validation.date.required')),
    genres: yup.array().min(1, t('addModal.validation.genre')),
    overview: yup.string().required(t('addModal.validation.overview')),
    runtime: yup
      .number()
      .required(t('addModal.validation.runtime.required'))
      .positive(t('addModal.validation.runtime.positive'))
      .integer(t('addModal.validation.runtime.integer')),
  })
  return (
    <Modal title={t('addModal.title')} onClose={onClose}>
      <Formik
        initialValues={{
          title: '',
          release_date: '',
          overview: '',
          runtime: '',
          genres: [],
          poster_path: '',
        }}
        onSubmit={(values) => {
          dispatch(addMovie(values))
          onClose(true)
        }}
        validationSchema={validationSchema}
      >
        {({ values, touched, errors, handleChange, handleSubmit, handleReset }) => {
          const { title, release_date, overview, runtime, genres, poster_path } = values
          return (
            <>
              <form className={styles.form}>
                <Input
                  title={t('addModal.input.title')}
                  name="title"
                  type="text"
                  value={title}
                  placeholder={t('addModal.placeholders.title')}
                />
                <br />
                <Input
                  title={t('addModal.input.date')}
                  name="release_date"
                  type="date"
                  value={release_date}
                  style={{ color: release_date ? '#fff' : '#a9a9a9c0' }}
                />
                <br />
                <Input
                  title={t('addModal.input.poster')}
                  name="poster_path"
                  type="text"
                  value={poster_path}
                  placeholder={t('addModal.placeholders.poster')}
                />
                <br />
                <FormSelect name="genres" value={genres.join(', ')} onChange={handleChange} />
                {touched.genres && errors.genres && <p className={styles.error}>{errors.genres}</p>}
                <br />
                <Input
                  title={t('addModal.input.overview')}
                  name="overview"
                  type="text"
                  value={overview}
                  placeholder={t('addModal.placeholders.overview')}
                />
                <br />
                <Input
                  title={t('addModal.input.runtime')}
                  name="runtime"
                  type="number"
                  value={runtime}
                  placeholder={t('addModal.placeholders.runtime')}
                />
              </form>
              <div className={styles.btns}>
                <Button onClick={handleReset} styleType={EStyleTypeBtn.reset}>
                  {t('addModal.reset')}
                </Button>
                <Button type={EType.submit} onClick={handleSubmit} styleType={EStyleTypeBtn.confirm}>
                  {t('addModal.submit')}
                </Button>
              </div>
            </>
          )
        }}
      </Formik>
    </Modal>
  )
}
