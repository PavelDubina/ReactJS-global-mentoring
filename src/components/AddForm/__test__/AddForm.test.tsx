import React from 'react'
import configureStore from 'redux-mock-store'
import ReactDOM from 'react-dom'
import '@testing-library/jest-dom/extend-expect'
import { render, fireEvent, waitFor } from '@testing-library/react'
import { Provider } from 'react-redux'
import movies from '../../../MockedData/mockMovies.json'
import { AddForm } from '../AddForm'

const mockStore = configureStore([])
jest.mock('next-i18next', () => ({
  useTranslation: jest.fn().mockReturnValue({
    t: (str: any) => str,
  }),
}))
describe('AddForm', () => {
  const store = mockStore({
    moviesData: movies,
  })
  ReactDOM.createPortal = jest.fn((element: any) => element)

  const getComponent = ({ onClose = jest.fn() }) =>
    render(
      <Provider store={store}>
        <AddForm onClose={onClose} />
      </Provider>,
    )
  it('Should renders correctly', () => {
    expect(getComponent({}).asFragment()).toMatchSnapshot()
  })
  it('Should show validation on blur', async () => {
    const { getByTestId } = getComponent({})
    const input = getByTestId('title')
    fireEvent.blur(input)
    await waitFor(() => {
      expect(getByTestId('error')).not.toBe(null)
      expect(getByTestId('error')).toHaveTextContent('addModal.validation.title')
    })
  })
})
