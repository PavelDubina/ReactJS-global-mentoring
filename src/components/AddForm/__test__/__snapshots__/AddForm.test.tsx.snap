// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddForm Should renders correctly 1`] = `
<DocumentFragment>
  <div
    class="modal_backdrop"
  >
    <div
      class="modal_window"
    >
      <div
        class="close_btn"
      >
        ×
      </div>
      <h1
        class="title"
      >
        addModal.title
      </h1>
      <form
        class="form"
      >
        <label
          class="label"
        >
          addModal.input.title
          <input
            autocomplete="off"
            class="input"
            data-testid="title"
            name="title"
            placeholder="addModal.placeholders.title"
            type="text"
            value=""
          />
        </label>
        <br />
        <label
          class="label"
        >
          addModal.input.date
          <input
            autocomplete="off"
            class="input"
            data-testid="release_date"
            name="release_date"
            style="color: rgba(169, 169, 169, 0.753);"
            type="date"
            value=""
          />
        </label>
        <br />
        <label
          class="label"
        >
          addModal.input.poster
          <input
            autocomplete="off"
            class="input"
            data-testid="poster_path"
            name="poster_path"
            placeholder="addModal.placeholders.poster"
            type="text"
            value=""
          />
        </label>
        <br />
        <span>
          addModal.input.genre
        </span>
        <div
          class="container"
        >
          <div
            class="select"
          >
            <p
              class="value default_value"
            >
              addModal.placeholders.genre
            </p>
            <div
              class="arrow"
            >
              ▼
            </div>
          </div>
          <div
            class="dropdown"
            data-testid="dropdown"
          >
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="action"
                name="genres"
                type="checkbox"
                value="action"
              />
              <label
                class="lable"
                for="action"
              >
                <span>
                  addModal.genres.action
                </span>
              </label>
            </div>
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="adventure"
                name="genres"
                type="checkbox"
                value="adventure"
              />
              <label
                class="lable"
                for="adventure"
              >
                <span>
                  addModal.genres.adventure
                </span>
              </label>
            </div>
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="animation"
                name="genres"
                type="checkbox"
                value="animation"
              />
              <label
                class="lable"
                for="animation"
              >
                <span>
                  addModal.genres.animation
                </span>
              </label>
            </div>
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="comedy"
                name="genres"
                type="checkbox"
                value="comedy"
              />
              <label
                class="lable"
                for="comedy"
              >
                <span>
                  addModal.genres.comedy
                </span>
              </label>
            </div>
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="crime"
                name="genres"
                type="checkbox"
                value="crime"
              />
              <label
                class="lable"
                for="crime"
              >
                <span>
                  addModal.genres.crime
                </span>
              </label>
            </div>
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="documentary"
                name="genres"
                type="checkbox"
                value="documentary"
              />
              <label
                class="lable"
                for="documentary"
              >
                <span>
                  addModal.genres.documentary
                </span>
              </label>
            </div>
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="drama"
                name="genres"
                type="checkbox"
                value="drama"
              />
              <label
                class="lable"
                for="drama"
              >
                <span>
                  addModal.genres.drama
                </span>
              </label>
            </div>
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="family"
                name="genres"
                type="checkbox"
                value="family"
              />
              <label
                class="lable"
                for="family"
              >
                <span>
                  addModal.genres.family
                </span>
              </label>
            </div>
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="fantasy"
                name="genres"
                type="checkbox"
                value="fantasy"
              />
              <label
                class="lable"
                for="fantasy"
              >
                <span>
                  addModal.genres.fantasy
                </span>
              </label>
            </div>
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="history"
                name="genres"
                type="checkbox"
                value="history"
              />
              <label
                class="lable"
                for="history"
              >
                <span>
                  addModal.genres.history
                </span>
              </label>
            </div>
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="horror"
                name="genres"
                type="checkbox"
                value="horror"
              />
              <label
                class="lable"
                for="horror"
              >
                <span>
                  addModal.genres.horror
                </span>
              </label>
            </div>
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="music"
                name="genres"
                type="checkbox"
                value="music"
              />
              <label
                class="lable"
                for="music"
              >
                <span>
                  addModal.genres.music
                </span>
              </label>
            </div>
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="mystery"
                name="genres"
                type="checkbox"
                value="mystery"
              />
              <label
                class="lable"
                for="mystery"
              >
                <span>
                  addModal.genres.mystery
                </span>
              </label>
            </div>
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="romance"
                name="genres"
                type="checkbox"
                value="romance"
              />
              <label
                class="lable"
                for="romance"
              >
                <span>
                  addModal.genres.romance
                </span>
              </label>
            </div>
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="science fiction"
                name="genres"
                type="checkbox"
                value="science fiction"
              />
              <label
                class="lable"
                for="science fiction"
              >
                <span>
                  addModal.genres.science fiction
                </span>
              </label>
            </div>
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="thriller"
                name="genres"
                type="checkbox"
                value="thriller"
              />
              <label
                class="lable"
                for="thriller"
              >
                <span>
                  addModal.genres.thriller
                </span>
              </label>
            </div>
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="war"
                name="genres"
                type="checkbox"
                value="war"
              />
              <label
                class="lable"
                for="war"
              >
                <span>
                  addModal.genres.war
                </span>
              </label>
            </div>
            <div
              class="input"
            >
              <input
                class="checkbox"
                id="western"
                name="genres"
                type="checkbox"
                value="western"
              />
              <label
                class="lable"
                for="western"
              >
                <span>
                  addModal.genres.western
                </span>
              </label>
            </div>
          </div>
        </div>
        <br />
        <label
          class="label"
        >
          addModal.input.overview
          <input
            autocomplete="off"
            class="input"
            data-testid="overview"
            name="overview"
            placeholder="addModal.placeholders.overview"
            type="text"
            value=""
          />
        </label>
        <br />
        <label
          class="label"
        >
          addModal.input.runtime
          <input
            autocomplete="off"
            class="input"
            data-testid="runtime"
            name="runtime"
            placeholder="addModal.placeholders.runtime"
            type="number"
            value=""
          />
        </label>
      </form>
      <div
        class="btns"
      >
        <button
          class="reset"
          type="button"
        >
          addModal.reset
        </button>
        <button
          class="confirm"
          type="submit"
        >
          addModal.submit
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
